### PR TITLE
Add 1D NURBS [1D-NURBS-dev]

### PR DESCRIPTION
**Pull request by IdoAkkerman**
This PR adds 1D NURBS meshes.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3529](https://github.com/mfem/mfem/pull/3529) | @IdoAkkerman | @tzanio | @dylan-copeland + @tzanio | 3/24/23 | 4/9/23 | 4/14/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by IdoAkkerman** [comment]
TODO - Boundary stuff:
 - Neumann and Robin BCs
 - Periodic BCs

**Comment by IdoAkkerman** [comment]
Are the boundary integrals sufficiently covered with the weak BC case? Or is it useful to also include non-homogenous neumann and/or robin BCs?

File: fem/fe_coll.cpp [hunk 1]
```diff
@@ -3468,16 +3469,18 @@ void NURBSFECollection::SetOrder(int Order) const
...
+   delete QuadrilateralFE;
+   delete ParallelepipedFE;
 }
 
 const FiniteElement *
 NURBSFECollection::FiniteElementForGeometry(Geometry::Type GeomType) const
 {
    switch (GeomType)
    {
+      case Geometry::POINT:       return PointFE;
```
Comment (dylan-copeland) [inline fem/fe_coll.cpp:3483]: Is `PointFE` ever used, or is it just added here for completeness?

File: miniapps/nurbs/nurbs_ex1.cpp [hunk 2]
```diff
@@ -10,6 +10,7 @@
 //               nurbs_ex1 -m ../../data/disc-nurbs.mesh -o -1
 //               nurbs_ex1 -m ../../data/pipe-nurbs.mesh -o -1
 //               nurbs_ex1 -m ../../data/beam-hex-nurbs.mesh -pm 1 -ps 2
+//               nurbs_ex1 -m ../../data/segment-nurbs.mesh -r 2 -o 2 -lod 3
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_ex1.cpp:13]: Does glvis work on the 1D output if glvis is rebuilt with this branch?

File: mesh/nurbs.cpp [hunk 3]
```diff
@@ -1821,6 +1866,84 @@ void NURBSExtension::ConnectBoundaries()
    GenerateBdrElementDofTable();
 }
 
+void NURBSExtension::ConnectBoundaries1D(int bnd0, int bnd1)
+{
+   /*  int idx0 = -1, idx1 = -1;
```
Comment (dylan-copeland) [inline mesh/nurbs.cpp:1871]: Remove commented code?

File: fem/fe_coll.cpp (hunk 1 above)
Comment (IdoAkkerman) [reply fem/fe_coll.cpp:3483]: I think it is used as boundary of a 1D mesh. Not 100% sure though...

File: miniapps/nurbs/nurbs_ex1.cpp (hunk 2 above)
Comment (IdoAkkerman) [reply miniapps/nurbs/nurbs_ex1.cpp:13]: Have not tried. I usually use visit for viz. I will check glvis.

File: fem/fe_coll.cpp (hunk 1 above)
Comment (IdoAkkerman) [reply fem/fe_coll.cpp:3483]: Checked it and it is indeed necessary for the 1D BCs.

File: mesh/nurbs.cpp (hunk 3 above)
Comment (IdoAkkerman) [reply mesh/nurbs.cpp:1871]: I am working on 1D periodicity. I will post a small code changes later today.

File: miniapps/nurbs/nurbs_ex1.cpp (hunk 2 above)
Comment (IdoAkkerman) [reply miniapps/nurbs/nurbs_ex1.cpp:13]: No unfortunately not. Apparently Glvis visualizes a 1D solution as a 2D solution  by extruding it.
This extrusion currently does not work for NURBS. Simply adding a NURBS FE to the extrusion  routine (around line 12350 in mesh.cpp) does not work, as NURBS spaces are constructed differently when compared to most other FEs. Therefore it is non-trivial to make the extrusion member functions of the Mesh class work for NURBS.

File: mesh/nurbs.cpp (hunk 3 above)
Comment (IdoAkkerman) [reply mesh/nurbs.cpp:1871]: There are still some other issues with periodic BCs. That are being addressed in #2058

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]
Thanks for adding this @IdoAkkerman !

**Comment by tzanio** [comment]
Should we mention it in `CHANGELOG`?

File: mesh/nurbs.hpp [hunk 4]
```diff
@@ -95,14 +95,16 @@ class NURBSPatch
...
 
    void swap(NURBSPatch *np);
 
    // Special B-NET access functions
    int SetLoopDirection(int dir);
-   inline       double &operator()(int i, int j);
-   inline const double &operator()(int i, int j) const;
+   inline       double &slice(int i, int j);
+   inline const double &slice(int i, int j) const;
+
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:106]: 2 extra empty lines

File: mesh/nurbs.hpp [hunk 5]
```diff
@@ -95,14 +95,16 @@ class NURBSPatch
...
-   int sd, nd;
+   int sd, nd, ls;
 
    void swap(NURBSPatch *np);
 
    // Special B-NET access functions
    int SetLoopDirection(int dir);
-   inline       double &operator()(int i, int j);
-   inline const double &operator()(int i, int j) const;
+   inline       double &slice(int i, int j);
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:104]: Please add a comment to describe this new function `slice`. It would be helpful to mention how it is different from the `operator()` function, and how it works in different dimensions (1D, 2D, 3D).

File: mesh/nurbs.hpp [hunk 6]
```diff
@@ -95,14 +95,16 @@ class NURBSPatch
 
    Array<KnotVector *> kv;
 
-   int sd, nd;
+   int sd, nd, ls;
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:98]: These variables with short, undescriptive names should be documented.

File: mesh/nurbs.hpp [hunk 7]
```diff
@@ -485,16 +497,43 @@ class NURBSPatchMap
 
 // Inline function implementations
 
-inline double &NURBSPatch::operator()(int i, int j)
+inline double &NURBSPatch::slice(int i, int j)
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:500]: Should there be index checks in the case `MFEM_DEBUG`?

File: mesh/nurbs.hpp [hunk 8]
```diff
@@ -485,16 +497,43 @@ class NURBSPatchMap
...
 }
 
+
+inline double &NURBSPatch::operator()(int i, int l)
+{
+#ifdef MFEM_DEBUG
+   if (data == 0 || i < 0 || i >= ni || nj > 0 || nk > 0 ||
+       l < 0 || l >= Dim)
+   {
+      mfem_error("NURBSPatch::operator() 1D");
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:517]: The meaning of `1D` in the error message is unclear.

File: miniapps/nurbs/nurbs_ex1.cpp (hunk 2 above)
Comment (dylan-copeland) [reply miniapps/nurbs/nurbs_ex1.cpp:13]: This example outputs glvis and VisIt files, without any option to disable the output. Users may be confused that those files cannot be viewed. The output of these files could be disabled by a flag (possibly parsed from the user's input), which also verifies that the 1D output is not written.

File: mesh/nurbs.hpp (hunk 4 above)
Comment (IdoAkkerman) [reply mesh/nurbs.hpp:106]: Removed

File: mesh/nurbs.hpp (hunk 5 above)
Comment (IdoAkkerman) [reply mesh/nurbs.hpp:104]: The slice operator is actually identical to the old operator(). 

I needed to extend the  _Standard B-NET access functions_  as defined in lines 140 - 147 to include the 1D case.
For these direct ijk access to the B_net we use  operator(). 

As mentioned lines 102 - 105 define _Special B-NET access functions_. 
Effectively  the `SetLoopDirection(int dir)` direction flattens the multi-dimensional B-NET to get an effective 1D net that can be modified using 'InsertKnot', 'DegreeElevate' or 'UniformRefinement'. The slice operator is the access function in that flattened structure. The first int gives the slice and the second int the element in that slice.

Long story short:
Nothing changed here, except the name
I will mention this in the comment
I will try to give a short and clear explanation of what slice does...

File: mesh/nurbs.hpp (hunk 8 above)
Comment (IdoAkkerman) [reply mesh/nurbs.hpp:517]: The first int gives the location of the net, the second int the variable  (coord or weight).

In the 2D and 3D cases there are also n+1 arguments.

File: mesh/nurbs.hpp (hunk 7 above)
Comment (IdoAkkerman) [reply mesh/nurbs.hpp:500]: Indeed add a check on existence of data and the index limits.

File: mesh/nurbs.hpp (hunk 6 above)
Comment (IdoAkkerman) [reply mesh/nurbs.hpp:98]: Added explanation:
```
   int nd; // Number of knots in flattened structure
   int ls; // Number of variables per knot in flattaned structure
   int sd; // Stride for data access
```

File: mesh/nurbs.hpp [hunk 9]
```diff
@@ -95,18 +95,23 @@ class NURBSPatch
...
-
    // Special B-NET access functions
+   //  - SetLoopDirection(int dir) flattens the multi-dimensional B-NET in the
+   //    requested direction. It effectively creates a 1D net.
+   //  - The slice(int, int) operator is the access function in that flattened structure.
+   //    The first int gives the slice and the second int the element in that slice.
+   //  - Both routines are used in 'InsertKnot', 'DegreeElevate' and 'UniformRefinement'.
+   //  - In older implementations slice(int int) was implemented as operator()(int, int)
+   int nd; // Number of knots in flattened structure
+   int ls; // Number of variables per knot in flattaned structure
```
Comment (dylan-copeland) [inline mesh/nurbs.hpp:106]: Thanks for the documentation. One typo: `flattaned` -> `flattened`

File: mesh/nurbs.cpp [hunk 10]
```diff
@@ -3246,6 +3487,33 @@ void NURBSExtension::SetSolutionVector(Vector &coords, int vdim)
...
+void NURBSExtension::Set1DSolutionVector(Vector &coords, int vdim)
+{
+   Array<const KnotVector *> kv(1);
+   NURBSPatchMap p2g(this);
+
+   weights.SetSize(GetNDof());
+   for (int p = 0; p < GetNP(); p++)
+   {
+      p2g.SetPatchDofMap(p, kv);
+      NURBSPatch &Patch = *patches[p];
```
Comment (dylan-copeland) [inline mesh/nurbs.cpp:3499]: MFEM has (unwritten, unenforced?) style that variable names should begin with a lower-case letter: `Patch` -> `patch`

File: miniapps/nurbs/nurbs_ex1.cpp [hunk 11]
```diff
@@ -401,6 +405,34 @@ int main(int argc, char *argv[])
...
+      for (int i = 0; i <  mesh->GetNE(); i++)
+      {
+         int geom       = mesh->GetElementBaseGeometry(i);
+         RefinedGeometry *refined_geo = GlobGeometryRefiner.Refine(( Geometry::Type)geom,
+                                                                   lod, 1);
+
+         x.GetValues(i, refined_geo->RefPts, vals);
+         nodes->GetValues(i, refined_geo->RefPts, coords);
+
+         for (int j = 0; j < vals.Size(); j++)
```
Comment (dylan-copeland) [inline miniapps/nurbs/nurbs_ex1.cpp:429]: This output of solution values is better than no output, and it is not the fault of this PR that MFEM has no visualization tool yet for these 1D NURBS solutions. However, it is unfortunate that these solution values cannot be immediately visualized. Also, `solution.dat` has redundant points, at the endpoints of elements. A user needs to set up their own 1D visualization tool, which is not too difficult, but it would take a bit of work, and the redundant points need to be handled. Perhaps a brief comment would be helpful, to inform the user that there is currently no visualization tool in MFEM for this data, and that there are redundant points output at the element interfaces.

**Review by dylan-copeland — APPROVED** [review]
Thanks @IdoAkkerman for adding this new feature.

File: mesh/nurbs.hpp (hunk 9 above)
Comment (IdoAkkerman) [reply mesh/nurbs.hpp:106]: Thanks.

File: miniapps/nurbs/nurbs_ex1.cpp (hunk 11 above)
Comment (IdoAkkerman) [reply miniapps/nurbs/nurbs_ex1.cpp:442]: Visualization  is very easy, as it is 1D. Gnuplot will do...

I cleaned the output. It is sorted and made unique. 
Perhaps those few lines could be moved to Gridfunction?

**Comment by tzanio** [comment]
Merged in `next` for testing...
